### ISSUE TITLE
Different representation of programs before and after typechecking

### ISSUE THREAD

**Issue by TomasMikula** [unisonweb/unison#513]
Currently, the same types `Term` and `Type` are used to represent programs before and after typechecking. I propose to represent typechecked programs differently than programs before typechecking.

### Motivation

 1. Express at the type level that a term is typechecked. Consumers of typed programs would have static guarantees of them being typed, instead of keeping track of that in programmer's head.
 2. Better starting point for pluggable syntax and pluggable typechecker. Different typecheckers could use different term types on input. Now we are at risk that `Term` and `Type` accomodate peculiarities of the default syntax or the default typechecker (for example, `IntroOuter` constructor of `Type` seems like a peculiarity of Haskellish syntax).
3. Have a more "normalized" representation of typed programs to avoid anomalies mentioned below.

### Suggestion

I don't have it all thought out, but for typechecked terms I would suggest to have a required type annotation at each level of a typed term, and the type annotation would contribute to the hash.

Types of typechecked terms should not have the `IntroOuter` case. 

#### Terms differing just by an explicit type annotation are considered different

Given source

```
id1 : Nat -> Nat
id1 x = x

id2 : Nat -> Nat
id2 x = (x: Nat)

> id1 == id2
```

the two functions differ:

```
  ┌
  │  7 | > id1 == id2
  │        ⧩
  │        false
  └
```

If in typechecked terms there was a type annotation at each subterm, the two functions would be the same.

#### Terms with the same source but different types are considered equal

The source

```
idNat: Nat -> Nat
idNat i = i

idInt: Int -> Int
idInt i = i
```

is successfully typechecked

```
  ┌
  │  ✅
  │  
  │  I found and typechecked these definitions in /Users/thomas/projects/unison/source.u:
  │  
  │    idInt : Int -> Int
  │    idNat : Nat -> Nat
  │  
  │  Now evaluating any watch expressions (lines starting with `>`)...
  └
```

but only one of them can be added to the codebase, because they have the same hash:

```
master> add

  ┌
  │  ✅
  │  
  │  I added these definitions:
  │  
  │    idNat : Nat -> Nat
  └
```

If every sub-term had a type annotation that would contribute to the hash, the two functions would have different hashes.

**Comment by pchiusano** [comment]
Hmm, you should be able to add both idNat and idInt even with the current implementation. 

What we do now is try inferring the type of the binding without the annotation. If that matches the annotation, we drop it. If the annotation is a super type, the term plus its annotation is used to compute the hash. I’m guessing there’s just a bug with this.

**Comment by TomasMikula** [comment]
OK, shall I open a separate issue for that?

**Comment by pchiusano** [comment]
Yes that would be great. You can create separate issue to track that bug.

I’ll noodle on the other ideas a bit. In general, we haven’t really gone to heroic lengths to make all equivalent terms hash the same and that is actually fine. The role of the hashing is more to unambiguously identify something in an unchanging way; that it can catch some obvious duplicates is more of a nice bonus. 

No matter what the hashing algorithm, for any nontrivial function there will be lots of different phrasings of the same function which hash differently. That’s ok.

Related: it would be nice to have a discovery mechanism that tells you about definitions before you’ve finished writing a duplicate... 😀

**Comment by TomasMikula** [comment]
Yeah, I wouldn't even expect `x -> f x` to hash the same as `f`. Somehow I viewed type annotations differently, as just hints to the typechecker or the human reader. Anyway, I will happily discount the value of this bonus feature from the value of this proposal.

Here are some further notes about the proposal:
 - The typechecker gathers a lot of information. This information is then reduced to a single bit, typechecked or not typechecked. Everything else is thrown away.

   Analogy: Parse a string as JSON, then throw away all the JSON structure, keep the original string and remember that it was succesfully parsed as JSON.

   Storing type information of each subterm of a typed program would retain the learned information.

 - Code stored in the codebase crucially depends on type inference. Just changing the type inference algorithm could make terms stored in the codebase not type-checkable with the new algorithm, or their hash would change with a new type-inference algorithm (from your comment above it follows that a smarter type inference could make more type annotations redundant, and thus not part of the hash).

   It would be cleaner to be able to define the codebase format without referring to type inference. There would be one true representation of typed programs, independent of how any particular type checker constructs such typed programs.

 - Today, how would we implement a `view` command that shows the full definition, including type applications and explicit `forall`s? In some future GUI structural editor, how could we display the type of any chosen subexpression? AFAICT, you would have to run type inference again, because that information was not stored.
